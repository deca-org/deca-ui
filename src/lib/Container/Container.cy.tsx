import { Test } from '../Utils';
import _cyp from '../../../cypress';
import React from 'react';
import Container from './Container';
import Box from '../Box';

const ContainerItem = () => (
  <Box
    css={{
      bg: '$indigo500',
      textAlign: 'center',
      padding: '$3',
      color: '$white',
      br: '$xs',
    }}
  >
    Content
  </Box>
);

describe('components/Container', () => {
  const selector = '[data-testid="test.container"]';
  it('fluid', () => {
    cy.baseMount(
      <Container data-testid="test.container" fluid>
        <ContainerItem />
      </Container>
    );
    cy.get(selector).should('have.css', 'max-width', '100%');
  });
  describe('px', () => {
    it('none', () => {
      cy.baseMount(
        <Container data-testid="test.container" px="none">
          <ContainerItem />
        </Container>
      );
      cy.get(selector).should('have.css', 'padding-left', Test.space('n'));
      cy.get(selector).should('have.css', 'padding-right', Test.space('n'));
    });
    it('sm', () => {
      cy.baseMount(
        <Container data-testid="test.container" px="sm">
          <ContainerItem />
        </Container>
      );
      cy.get(selector).should('have.css', 'padding-left', Test.space('2'));
      cy.get(selector).should('have.css', 'padding-right', Test.space('2'));
    });
    it('md', () => {
      cy.baseMount(
        <Container data-testid="test.container" px="md">
          <ContainerItem />
        </Container>
      );
      cy.get(selector).should('have.css', 'padding-left', Test.space('3'));
      cy.get(selector).should('have.css', 'padding-right', Test.space('3'));
    });
    it('lg', () => {
      cy.baseMount(
        <Container data-testid="test.container" px="lg">
          <ContainerItem />
        </Container>
      );
      cy.get(selector).should('have.css', 'padding-left', Test.space('4'));
      cy.get(selector).should('have.css', 'padding-right', Test.space('4'));
    });
  });
  describe('breakpoints', () => {
    it('xs', () => {
      cy.viewport(300, 500);
      cy.baseMount(
        <Container data-testid="test.container">
          <ContainerItem />
        </Container>
      );
      cy.get(selector).should(
        'have.css',
        'max-width',
        Test.breakpoint('xs') + 'px'
      );
    });
    it('sm', () => {
      cy.viewport(Test.breakpoint('sm'), 500);
      cy.baseMount(
        <Container data-testid="test.container">
          <ContainerItem />
        </Container>
      );
      cy.get(selector).should(
        'have.css',
        'max-width',
        Test.breakpoint('sm') + 'px'
      );
    });
    it('md', () => {
      cy.viewport(Test.breakpoint('md'), 500);
      cy.baseMount(
        <Container data-testid="test.container">
          <ContainerItem />
        </Container>
      );
      cy.get(selector).should(
        'have.css',
        'max-width',
        Test.breakpoint('sm') + 'px'
      );
    });
    it('lg', () => {
      cy.viewport(Test.breakpoint('lg'), 500);
      cy.baseMount(
        <Container data-testid="test.container">
          <ContainerItem />
        </Container>
      );
      cy.get(selector).should(
        'have.css',
        'max-width',
        Test.breakpoint('md') + 'px'
      );
    });
    it('xl', () => {
      cy.viewport(Test.breakpoint('xl'), 500);
      cy.baseMount(
        <Container data-testid="test.container">
          <ContainerItem />
        </Container>
      );
      cy.get(selector).should(
        'have.css',
        'max-width',
        Test.breakpoint('lg') + 'px'
      );
    });
  });
});
