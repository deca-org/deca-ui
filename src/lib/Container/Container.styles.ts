import { styled } from '@lib/Theme/stitches.config';

export const StyledContainer = styled('div', {
  w: '100%',
  mr: 'auto',
  ml: 'auto',
  variants: {
    px: {
      none: {
        px: '$n',
      },
      sm: {
        px: '$2',
      },
      md: {
        px: '$3',
      },
      lg: {
        px: '$4',
      },
    },
    responsive: {
      true: {
        '@n': {
          maxWidth: '$breakpoints$xs',
        },
        '@xs': {
          maxWidth: '$breakpoints$xs',
        },
        '@sm': {
          maxWidth: '$breakpoints$sm',
        },
        '@md': {
          maxWidth: '$breakpoints$sm',
        },
        '@lg': {
          maxWidth: '$breakpoints$md',
        },
        '@xl': {
          maxWidth: '$breakpoints$lg',
        },
      },
    },
    fluid: {
      true: {
        maxWidth: '100%',
      },
    },
  },
});
