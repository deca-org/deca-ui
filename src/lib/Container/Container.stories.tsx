import Box from '@lib/Box';
import { ComponentMeta, ComponentStory } from '@storybook/react';
import React from 'react';

import Container from './Container';

export default {
  title: 'Container',
  component: Container,
} as ComponentMeta<typeof Container>;

const Template: ComponentStory<typeof Container> = (args) => (
  <Container {...args}>
    <Box
      css={{
        bg: '$indigo500',
        br: '$md',
        color: '$white',
        p: '$3',
      }}
    >
      Lorem ipsum dolor sit amet, consectetur adipiscing elit. Mauris ipsum
      justo, interdum eget lacus sit amet, consequat scelerisque lectus.
      Suspendisse vel tincidunt purus.
    </Box>
  </Container>
);

export const Responsive = Template.bind({});

Responsive.args = {
  responsive: true,
  px: 'sm',
  fluid: false,
};

export const Fluid = Template.bind({});

Fluid.args = {
  responsive: true,
  px: 'sm',
  fluid: true,
};
