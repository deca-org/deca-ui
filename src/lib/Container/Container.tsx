import { CSS } from '@lib/Theme/stitches.config';
import {
  PolymorphicRef,
  PolymorphicComponentPropWithRef,
  __DEV__,
} from '@lib/Utils';
import clsx from 'clsx';
import React from 'react';

import { StyledContainer } from './Container.styles';

/**
 * The Container component fixes an element's width to the current breakpoint
 */
interface Props {
  /**
   * Override default CSS style.
   */
  css?: CSS;
  /**
   * The content of the component.
   */
  children?: React.ReactNode | undefined;
  /**
   * ClassName applied to the component.
   * @default ''
   */
  className?: string;
  /**
   * padding applied on each side
   * @default sm
   */
  px?: 'none' | 'sm' | 'md' | 'lg';
  /**
   * container max-width changes with breakpoint
   * @default true
   */
  responsive?: boolean;
  /**
   * max-width is set to 100%
   */
  fluid?: boolean;
}

export type ContainerProps<T extends React.ElementType> =
  PolymorphicComponentPropWithRef<T, Props>;

export type ContainerComponent = (<C extends React.ElementType = 'div'>(
  props: ContainerProps<C>
) => React.ReactElement | null) & { displayName?: string };

const Container: ContainerComponent = React.forwardRef(
  <T extends React.ElementType = 'div'>(
    {
      as,
      css,
      className = '',
      children,
      px = 'sm',
      responsive = true,
      fluid = false,
      ...containerProps
    }: ContainerProps<T>,
    ref?: PolymorphicRef<T>
  ) => {
    const preClass = 'decaContainer';

    return (
      <StyledContainer
        as={as}
        css={css}
        className={clsx(className, `${preClass}-root`)}
        px={px}
        responsive={responsive}
        fluid={fluid}
        ref={ref}
        {...containerProps}
      >
        {children}
      </StyledContainer>
    );
  }
);

if (__DEV__) {
  Container.displayName = 'DecaUI.Container';
}

export default Container;
