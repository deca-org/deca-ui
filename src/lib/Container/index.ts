import Container from './Container';
export * from './Container';
export { Container };
export default Container;
