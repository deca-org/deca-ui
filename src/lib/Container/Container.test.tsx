import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import React from 'react';

import Container from './Container';

describe('components/Container', () => {
  it('renders content', () => {
    const { getByText } = render(<Container>Children</Container>);
    expect(getByText('Children')).toBeInTheDocument();
  });
  it('matches snapshot', () => {
    const { asFragment } = render(<Container></Container>);
    expect(asFragment()).toMatchSnapshot();
  });
});
