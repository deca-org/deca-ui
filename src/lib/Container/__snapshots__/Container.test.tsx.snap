// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Container matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="c-fdHeMm c-fdHeMm-jkLaoV-px-sm c-fdHeMm-hVxVNe-responsive-true decaContainer-root"
  />
</DocumentFragment>
`;
